duckduckgo/macos-browser
# Memory leak "fixes"

Task/Issue URL:
CC: @tomasstrba 

**Description**:

There's an issue where `UserScript` instances are being leaked by the OS. When you close a tab, it and everything it holds onto (including its web view) get deinitialized, but the user scripts do not. After opening dozens of tabs, the app's memory usage gets very large.

This PR includes a possible way to lessen this problem, but it was only chosen after looking at this problem quickly. We use user scripts on iOS and do not have the same problem, so there's almost certainly a better solution. This PR prevents scripts from leaking, but the object used to do so itself leaks – it's just much smaller than the scripts are, so memory usage climbs far slower.

I don't think there's anything on our side causing this to leak. I made a number of tweaks to our code without luck, and if you comment out `webView.configuration.userContentController.add(userScript: $0)` then they get deallocated fine, suggesting that it's the user content controller causing problems. If you look in the memory graph debugger then you'll also see that they're being referenced by some WebKit related object.

**Steps to test this PR**:

1. Launch the app and open a number of tabs
2. Check how many instances of the user scripts are live in the Memory Graph Debugger; it should be equal to the number of tabs
3. Close all tabs except one
4. Check the user scripts in memory again, there should only be one of each

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
